pkg/image/admission/imagepolicy/imagepolicy.go

```diff
@@ -0,0 +1,392 @@
+
+			// if we should resolve the image, update the image and also check the cache again
+			if resolvesImage {
+				container.Image = ref.Exact()
+				isDigest = true
+			}
+		}
+	}
+
+	attr := &rules.ImagePolicyAttributes{
+		Name:         ref,
+		OriginalName: name,
+		Resource:     gr,
+		Image:        image,
+	}
+	// remember under both names
+	covered.Remember(name, attr)
+	covered.Remember(container.Image, attr)
+
+	accepted := a.accepter.Accepts(attr)
+
+	glog.V(5).Infof("Made decision for %v: %t", ref, accepted)
+	return accepted, nil
+}
+
+func (a *imagePolicyPlugin) adjust(podSpec *kapi.PodSpec, attr admission.Attributes, covered policyAttributes) error {
+	gr := attr.GetResource().GroupResource()
+	requiresImage := a.adjuster.RequiresImage(gr)
+
+	for i := range podSpec.Containers {
+		if err := a.adjustContainer(podSpec, &podSpec.Containers[i], gr, covered, requiresImage); err != nil {
+			return err
+		}
+	}
+	for i := range podSpec.InitContainers {
+		if err := a.adjustContainer(podSpec, &podSpec.InitContainers[i], gr, covered, requiresImage); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (a *imagePolicyPlugin) adjustContainer(podSpec *kapi.PodSpec, container *kapi.Container, gr unversioned.GroupResource, covered policyAttributes, requiresImage bool) error {
+	name := container.Image
+
+	// ignore containers without images
+	if len(name) == 0 {
+		return nil
+	}
+
+	// resolve the image if we have not done so - this logic is simpler than accept because we only need
+	// the image
+	attr, ok := covered.Get(name)
+	if !ok {
+		ref, err := imageapi.ParseDockerImageReference(name)
+		if err != nil {
+			return err
+		}
+
+		var image *imageapi.Image
+		if requiresImage {
+			image, err = a.resolver.Resolve(ref)
+			if err != nil {
+				return err
+			}
+			glog.V(5).Infof("Resolved image %v to %s (%s)", ref, image.Name, image.DockerImageReference)
+		}
+
+		attr = &rules.ImagePolicyAttributes{
+			Name:         ref,
+			OriginalName: name,
+			Resource:     gr,
+			Image:        image,
+		}
+		covered.Remember(name, attr)
+	}
+
+	a.adjuster.Adjust(attr, podSpec)
+	return nil
+}
+
+type policyAttributes map[string]*rules.ImagePolicyAttributes
+
+func (d policyAttributes) Remember(image string, attr *rules.ImagePolicyAttributes) {
+	d[image] = attr
+}
+
+func (d policyAttributes) Has(image string) bool {
+	_, ok := d[image]
+	return ok
+}
+
+func (d policyAttributes) Get(image string) (*rules.ImagePolicyAttributes, bool) {
+	attr, ok := d[image]
+	return attr, ok
+}
+
+type imageResolutionCache struct {
+	images     client.ImageInterface
+	tags       client.ImageStreamTagsNamespacer
+	integrated rules.RegistryMatcher
+	expiration time.Duration
+
+	cache *lru.Cache
+}
+
+type imageCacheEntry struct {
+	expires time.Time
+	image   *imageapi.Image
+}
+
+func newImageResolutionCache(images client.ImageInterface, tags client.ImageStreamTagsNamespacer, integratedRegistry rules.RegistryMatcher) (*imageResolutionCache, error) {
```

Comment:
I took a look at this and because we don't have shared code for IS -> ISTag I'm not going to do it - there's too much code movement and duplication (an IST is a delicate and unique flower that is complex to set up).  Will spawn an issue for that.